### PR TITLE
Backport ecAudit to Cassandra 2.2

### LINKED ISSUE #111: Backport ecAudit to 2.2.x
Potential users have requested a backport of ecAudit that would be compatible with Cassandra 2.2.8. We should investigate if this is possible, with our without compromises in supported features.

Ideally we should have the very same features on all flavors of ecAudit, if nothing else for sake of maintainability.

Also, since 2.2.8 is a rather old release of the 2.2.x series, we should aim to start with a backport to 2.2.14 (latest at the time of writing). If necessary we'd make a backport further to 2.2.8 if necessary.

### PR DESCRIPTION
**Pull request by eperott**
Closes #111

### CODE REVIEW

File: common/pom.xml [hunk 1]
```diff
@@ -19,10 +19,10 @@
     <modelVersion>4.0.0</modelVersion>
     <parent>
         <groupId>com.ericsson.bss.cassandra.ecaudit</groupId>
-        <artifactId>parent_c3.0.11</artifactId>
-        <version>2.1.0</version>
+        <artifactId>parent_c2.2.14</artifactId>
```
Comment (eperott) [inline common/pom.xml]: Should be parent_c2.2. Same applies for all artifactId's

File: ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/auth/AuditPasswordAuthenticator.java [hunk 2]
```diff
@@ -88,10 +87,10 @@ public void setup()
     }
 
     @Override
-    public SaslNegotiator newSaslNegotiator(InetAddress clientAddress)
+    public SaslNegotiator newSaslNegotiator()
     {
-        LOG.debug("Setting up SASL negotiation with {}", clientAddress);
-        return new AuditPlainTextSaslAuthenticator(clientAddress, wrappedAuthenticator.newSaslNegotiator(clientAddress));
+        LOG.debug("Setting up SASL negotiation with client peer");
+        return new AuditPlainTextSaslAuthenticator(wrappedAuthenticator.newSaslNegotiator());
```
Comment (eperott) [inline ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/auth/AuditPasswordAuthenticator.java]: We should consider to do some research on whether it is possible to extract client address some how without being to hacky.

File: ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/entry/PreparedAuditOperation.java [hunk 3]
```diff
@@ -88,7 +94,28 @@ private String preparedWithValues()
...
+                        SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.getDefault(Locale.Category.FORMAT));
+                        sdf.setTimeZone(TimeZone.getTimeZone("UTC"));
+                        value = sdf.format(TimestampSerializer.instance.deserialize(serializedValue));
+                    }
+                    else
+                    {
+                        value = column.type.getString(serializedValue);
+                    }
+                }
+            }
```
Comment (eperott) [inline ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/entry/PreparedAuditOperation.java]: This block should be refactored and preferably placed in a separate class.

File: ecaudit/src/test/java/com/ericsson/bss/cassandra/ecaudit/TestAuditAdapter.java [hunk 4]
```diff
@@ -121,7 +121,7 @@
     public static void beforeAll()
     {
         ClientInitializer.beforeClass();
-        oldPartitionerToRestore = DatabaseDescriptor.setPartitionerUnsafe(Mockito.mock(IPartitioner.class));
+        //oldPartitionerToRestore = DatabaseDescriptor.setPartitionerUnsafe(Mockito.mock(IPartitioner.class));
```
Comment (eperott) [inline ecaudit/src/test/java/com/ericsson/bss/cassandra/ecaudit/TestAuditAdapter.java]: Remove

File: ecaudit/src/test/java/com/ericsson/bss/cassandra/ecaudit/TestAuditAdapter.java [hunk 5]
```diff
@@ -143,7 +143,7 @@ public void after()
     @AfterClass
     public static void afterAll()
     {
-        DatabaseDescriptor.setPartitionerUnsafe(oldPartitionerToRestore);
+        //DatabaseDescriptor.setPartitionerUnsafe(oldPartitionerToRestore);
```
Comment (eperott) [inline ecaudit/src/test/java/com/ericsson/bss/cassandra/ecaudit/TestAuditAdapter.java]: Remove

File: ecaudit/src/test/java/com/ericsson/bss/cassandra/ecaudit/handler/TestAuditQueryHandler.java [hunk 6]
```diff
@@ -96,13 +96,13 @@
     public static void beforeAll()
     {
         ClientInitializer.beforeClass();
-        oldPartitionerToRestore = DatabaseDescriptor.setPartitionerUnsafe(Mockito.mock(IPartitioner.class));
+        //oldPartitionerToRestore = DatabaseDescriptor.setPartitionerUnsafe(Mockito.mock(IPartitioner.class));
```
Comment (eperott) [inline ecaudit/src/test/java/com/ericsson/bss/cassandra/ecaudit/handler/TestAuditQueryHandler.java]: Remove

File: ecaudit/src/test/java/com/ericsson/bss/cassandra/ecaudit/handler/TestAuditQueryHandler.java [hunk 7]
```diff
@@ -96,13 +96,13 @@
...
     }
 
     @AfterClass
     public static void afterAll()
     {
-        DatabaseDescriptor.setPartitionerUnsafe(oldPartitionerToRestore);
+        //DatabaseDescriptor.setPartitionerUnsafe(oldPartitionerToRestore);
```
Comment (eperott) [inline ecaudit/src/test/java/com/ericsson/bss/cassandra/ecaudit/handler/TestAuditQueryHandler.java]: Remove

File: integration-test-standard/src/test/java/com/ericsson/bss/cassandra/ecaudit/integration/custom/ITVerifyPostLogging.java [hunk 8]
```diff
@@ -176,34 +177,70 @@ private String replaceBatchIdWithConstant(String s)
     }
 
     @Test
-    public void testFailedStatements()
+    public void testFailedSimpleStatements()
```
Comment (eperott) [inline integration-test-standard/src/test/java/com/ericsson/bss/cassandra/ecaudit/integration/custom/ITVerifyPostLogging.java]: The idea here is that the original test case "testFailedStatements" is divided in 3 parts, where only the first part is applicable in the setup we have in 2.2. As we merge this into 3.0.11 we should remove the @Ignore on the two remainging test cases.

File: pom.xml [hunk 9]
```diff
@@ -87,6 +87,12 @@
 
     <dependencyManagement>
         <dependencies>
+            <dependency>
```
Comment (eperott) [inline pom.xml]: Euhm, don't remember why I added this. We should be getting it as a transitive dependency. But might be a way to resolve conflict between driver and embedded Cassandra. Find out and add comment?

File: test-utils/src/main/java/com/ericsson/bss/cassandra/ecaudit/test/daemon/CassandraDaemonForAuditTest.java [hunk 10]
```diff
@@ -73,6 +73,7 @@ private CassandraDaemonForAuditTest() throws IOException
     {
         setupConfiguration();
         cassandraDaemon = new CassandraDaemon(true);
+        //cassandraDaemon.init(null);
```
Comment (eperott) [inline test-utils/src/main/java/com/ericsson/bss/cassandra/ecaudit/test/daemon/CassandraDaemonForAuditTest.java]: Remove

File: ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/auth/AuditPasswordAuthenticator.java (hunk 2 above)
Comment (etedpet) [reply ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/auth/AuditPasswordAuthenticator.java]: Well, this method is called by the ServerConnection class, which has access to the ClientState which in turn has the address... 
BUT there is no way in Java (as far as I know) to get access to the calling instance from somewhere in the code. Even if we know which class that is calling us, we still need the instance to be able to use reflection to access that instance member data. (same goes for the AuditPlainTextSaslAuthenticator#getAuthenticatedUser() method below)
So I don't think it can be done...

https://stackoverflow.com/questions/43072542/can-i-get-the-instance-of-the-calling-object-in-java
Basically says:
"Short answer: you can't.
Long answer: you can get it if you either run it on a debugger, or your own custom JVM."

File: ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/entry/PreparedAuditOperation.java (hunk 3 above)
Comment (etedpet) [reply ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/entry/PreparedAuditOperation.java]: Added ColumnDataPrinter util class...
While testing that a new test dependency was added - JUnitParams, which enables data providers to tests... This is not used in the c3+ code for now, but maybe that can be useful for other tests in the future?

File: ecaudit/src/test/java/com/ericsson/bss/cassandra/ecaudit/TestAuditAdapter.java (hunk 4 above)
Comment (etedpet) [reply ecaudit/src/test/java/com/ericsson/bss/cassandra/ecaudit/TestAuditAdapter.java]: done

File: ecaudit/src/test/java/com/ericsson/bss/cassandra/ecaudit/TestAuditAdapter.java (hunk 5 above)
Comment (etedpet) [reply ecaudit/src/test/java/com/ericsson/bss/cassandra/ecaudit/TestAuditAdapter.java]: done

File: pom.xml (hunk 9 above)
Comment (etedpet) [reply pom.xml]: Removed this dependency and all tests still pass... And there is no problem running ecAudit in CCM either... So I guess we can remove it.

Comment (eperott) [reply pom.xml]: Agree

File: ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/utils/c2_2/ColumnDataPrinter.java [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+
+    private ColumnDataPrinter()
+    {
+        // Utility class
+    }
+
+    public static String toCQLLiteral(ByteBuffer serializedValue, ColumnSpecification column)
+    {
+        String value = null;
+        if (column.type instanceof UTF8Type || column.type instanceof AsciiType)
```
Comment (eperott) [inline ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/utils/c2_2/ColumnDataPrinter.java]: Break out condition into a method?
E.g: isStringType(type)

File: ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/utils/c2_2/ColumnDataPrinter.java [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+    private static final DateFormat DATE_FORMAT = createDateFormat();
+
+    private ColumnDataPrinter()
+    {
+        // Utility class
+    }
+
+    public static String toCQLLiteral(ByteBuffer serializedValue, ColumnSpecification column)
+    {
+        String value = null;
```
Comment (eperott) [inline ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/utils/c2_2/ColumnDataPrinter.java]: Would it be less tangled if we just return on each place rather than store in "value"?

File: ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/utils/c2_2/ColumnDataPrinter.java [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+        if (column.type instanceof UTF8Type || column.type instanceof AsciiType)
+        {
+            value = "'" + column.type.getString(serializedValue) + "'";
+        }
+        else
+        {
+            if (serializedValue.hasRemaining())
```
Comment (eperott) [inline ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/utils/c2_2/ColumnDataPrinter.java]: Hmm, do this check in begining of method and bail out with null result if false...

File: ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/utils/c2_2/ColumnDataPrinter.java (hunk 12 above)
Comment (etedpet) [reply ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/utils/c2_2/ColumnDataPrinter.java]: Changed to using multiple returns. Some people strongly dislikes that... But I'm fine with it, as long as the method is clean. Better than having a "deep logic" like before.

File: ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/utils/c2_2/ColumnDataPrinter.java (hunk 11 above)
Comment (etedpet) [reply ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/utils/c2_2/ColumnDataPrinter.java]: done

File: ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/utils/c2_2/ColumnDataPrinter.java (hunk 13 above)
Comment (etedpet) [reply ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/utils/c2_2/ColumnDataPrinter.java]: I thought we (you) wanted an non-existing string value to return '' :-)
But this gives cleaner code and better logic I think. 
Done!

File: ecaudit/src/test/java/com/ericsson/bss/cassandra/ecaudit/handler/TestAuditQueryHandler.java (hunk 7 above)
Comment (etedpet) [reply ecaudit/src/test/java/com/ericsson/bss/cassandra/ecaudit/handler/TestAuditQueryHandler.java]: done (again?) don't know if something strange happened with the merge or if I missed this the first time...

File: README.md [hunk 14]
```diff
@@ -38,6 +38,7 @@ Get the ecAudit flavor for your Cassandra version.
 [![ecAudit for Cassandra 3.11.<latest>](https://img.shields.io/maven-central/v/com.ericsson.bss.cassandra.ecaudit/ecaudit_c3.11.svg?label=ecAudit%20for%20Cassandra%203.11.<latest>)](https://search.maven.org/search?q=g:%22com.ericsson.bss.cassandra.ecaudit%22%20AND%20a:%22ecaudit_c3.11%22)
 [![ecAudit for Cassandra 3.0.<latest>](https://img.shields.io/maven-central/v/com.ericsson.bss.cassandra.ecaudit/ecaudit_c3.0.svg?label=ecAudit%20for%20Cassandra%203.0.<latest>)](https://search.maven.org/search?q=g:%22com.ericsson.bss.cassandra.ecaudit%22%20AND%20a:%22ecaudit_c3.0%22)
 [![ecAudit for Cassandra 3.0.11](https://img.shields.io/maven-central/v/com.ericsson.bss.cassandra.ecaudit/ecaudit_c3.0.11.svg?label=ecAudit%20for%20Cassandra%203.0.11)](https://search.maven.org/search?q=g:%22com.ericsson.bss.cassandra.ecaudit%22%20AND%20a:%22ecaudit_c3.0.11%22)
+[![ecAudit for Cassandra 2.2.<latest>](https://img.shields.io/maven-central/v/com.ericsson.bss.cassandra.ecaudit/ecaudit_c3.0.11.svg?label=ecAudit%20for%20Cassandra%202.2)](https://search.maven.org/search?q=g:%22com.ericsson.bss.cassandra.ecaudit%22%20AND%20a:%22ecaudit_c2.2%22)
```
Comment (eperott) [inline README.md]: Change shield link to 2.2. Will work once things are merged...

Comment (etedpet) [reply README.md]: done
